### PR TITLE
Pagination implemented in erlcloud_ec2

### PR DESCRIPTION
**Pull request by SergeyOvchinnik**
Problem:
7 erlcloud_ec2 describe functions with pagination implemented on API level did not support pagination feature on library level

Functions:
- describe_instances
- describe_instance_status
- describe_reserved_instance_offerings
- describe_snapshots
- describe_spot_fleet_instances
- describe_spot_price_history
- describe_tags

Solution:
Implemented pagination feature for all 7 functions
Implemented unit tests for pagination in those functions

Reviewers:
@ddosia @motobob

### CODE REVIEW

File: src/erlcloud_ec2.erl [hunk 1]
```diff
@@ -1348,37 +1352,73 @@ attribute_xpath(_, AttributeName) ->
     "/DescribeInstanceAttributeResponse/" ++ AttributeName ++ "/value".
 
 
+-type ec2_instances_ids() :: [string()].
+-type ec2_token() :: string() | undefined.
+-type ec2_max_result() :: pos_integer() | undefined.
+-type ok_error(Ok) :: {ok, Ok} | {error, term()}.
+-type ok_error(Ok1, Ok2) :: {ok, Ok1, Ok2} | {error, term()}.
```
Comment (motobob) [inline src/erlcloud_ec2.erl]: this seems to be 

``` erlang
ok_error(Ok, ec2_token()) :: {ok, Ok, ec2_token()} | {error, term()}
```

File: src/erlcloud_ec2.erl [hunk 2]
```diff
@@ -1441,25 +1481,58 @@ extract_block_device_mapping_status(Node) ->
...
+    when is_list(Params), is_list(Filter), is_record(Config, aws_config) ->
     AllParams = Params ++ list_to_ec2_filter(Filter),
     case ec2_query(Config, "DescribeInstanceStatus", AllParams, ?NEW_API_VERSION) of
         {ok, Doc} ->
-            Path = "/DescribeInstanceStatusResponse/instanceStatusSet/item",
-            {ok, [ extract_instance_status(Item) || Item <- xmerl_xpath:string(Path, Doc) ]};
-        {error, _} = Error ->
-            Error
+            Statuses =
+                extract_results("DescribeInstanceStatusResponse", "instanceStatusSet", fun extract_instance_status/1, Doc),
```
Comment (motobob) [inline src/erlcloud_ec2.erl]: please be consistent with line wrapping. pick one and use it everywhere

File: src/erlcloud_ec2.erl [hunk 3]
```diff
@@ -1713,43 +1786,81 @@ extract_reserved_instance(Node) ->
...
+-spec describe_reserved_instances_offerings(ec2_selector(), aws_config())
+    -> ok_error(proplist()) | ok_error(proplist(), ec2_token()).
 describe_reserved_instances_offerings(Selector, Config)
-  when is_list(Selector) ->
-    InstanceTypes = [Value || {Key, Value} <- Selector, Key =:= instance_type],
-    AvailabilityZones = [Value || {Key, Value} <- Selector, Key =:= availability_zone],
-    Descs = [Value || {Key, Value} <- Selector, Key =:= product_description],
-    Token = [Value || {Key, Value} <- Selector, Key =:= token],
+    when is_list(Selector), is_record(Config, aws_config) ->
+    InstanceTypes = proplists:get_value(instance_type, Selector),
```
Comment (motobob) [inline src/erlcloud_ec2.erl]: technically speaking previous implementation allowed several `instance_type`s  or rest of the filters to be passed.
Please check with AWS API doc and make sure it's implemented accordingly

File: src/erlcloud_ec2.erl [hunk 4]
```diff
@@ -1713,43 +1786,81 @@ extract_reserved_instance(Node) ->
...
+
+-spec describe_reserved_instances_offerings(ec2_selector(), ec2_max_result(), ec2_token(), aws_config())
+    -> ok_error(proplist(), ec2_token()).
+describe_reserved_instances_offerings(Selector, MaxResults, NextToken, Config)
+    when is_list(Selector), is_integer(MaxResults),
+         MaxResults >= ?RESERVED_INSTANCES_OFFERINGS_MR_MIN,
+         MaxResults =< ?RESERVED_INSTANCES_OFFERINGS_MR_MAX,
+         is_list(NextToken) orelse NextToken =:= undefined,
+         is_record(Config, aws_config) ->
+    InstanceTypes = proplists:get_value(instance_type, Selector),
```
Comment (motobob) [inline src/erlcloud_ec2.erl]: same applies here.

**Review by motobob — CHANGES_REQUESTED** [review]
several minor changes and it is finally good to go.

File: src/erlcloud_ec2.erl (hunk 1 above)
Comment (SergeyOvchinnik) [reply src/erlcloud_ec2.erl]: Well, `ok_error` is always parametrized as either `ok_error(proplist())` or `ok_error(proplist(), ec2_token())`, so I could just roll back to my previous implementation and not parametrize them at all. Doesn't make much difference for dialyzer anyway.

File: src/erlcloud_ec2.erl (hunk 3 above)
Comment (SergeyOvchinnik) [reply src/erlcloud_ec2.erl]: True, should have used `get_all_values` instead

File: src/erlcloud_ec2.erl (hunk 2 above)
Comment (SergeyOvchinnik) [reply src/erlcloud_ec2.erl]: fixed

File: src/erlcloud_ec2.erl (hunk 3 above)
Comment (SergeyOvchinnik) [reply src/erlcloud_ec2.erl]: fixed

File: src/erlcloud_ec2.erl (hunk 4 above)
Comment (SergeyOvchinnik) [reply src/erlcloud_ec2.erl]: fixed

**Review by motobob — APPROVED** [review]
I'm out of comments. let's 🚢 :shipit: 
@ddosia are you fine?

**Comment by motobob** [comment]
i actually wanted this to happen https://github.com/erlcloud/erlcloud/compare/master...SergeyOvchinnik:sovchinnik-erlcloud-ec2-pagination-feature but that's fine. we will merge it separatelly after

File: src/erlcloud_ec2.erl [hunk 5]
```diff
@@ -1441,25 +1481,57 @@ extract_block_device_mapping_status(Node) ->
...
 
-%%
-%%
--spec describe_instance_status(string()) -> proplist().
-describe_instance_status(InstanceID) ->
+-type instance_id() :: string().
+-type ec2_param_list() :: [{string(),string()}].
+
+-define(SPOT_INSTANCE_STATUS_MR_MIN, 5).
+-define(SPOT_INSTANCE_STATUS_MR_MAX, 1000).
```
Comment (ddosia) [inline src/erlcloud_ec2.erl]: Usually all "define"s are located somewhere in the top of the file.

**Review by ddosia — APPROVED** [review]
+1

**Comment by SergeyOvchinnik** [comment]
@ddosia I moved all defines and mutual types to the top, but defines don't look good at all. How would I optimize the so that the range is 5-1000 everywhere except the one place where it's 1-1000?

**Pull request merged**